components/tikv_util/src/metrics/memory_linux.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+//! This module is a subset of rust-prometheus's process collector, without the fd collector
+//! to avoid memory fragmentation issues when open fd is large.
+
+use std::io::{Error, ErrorKind, Result};
+
+use libc;
+use procinfo::pid as pid_info;
+
+use prometheus::core::{Collector, Desc};
+use prometheus::{proto, Gauge, Opts};
+
+/// Monitors threads of the current process.
+pub fn monitor_memory() -> Result<()> {
+    let pid = unsafe { libc::getpid() };
+    let tc = MemoryCollector::new(pid);
+    prometheus::register(Box::new(tc)).map_err(|e| Error::new(ErrorKind::Other, e.to_string()))
+}
+
+/// A collector to collect memory metrics.
+pub struct MemoryCollector {
+    pid: libc::pid_t,
+    descs: Vec<Desc>,
+    vsize: Gauge,
+    rss: Gauge,
+}
+
+impl MemoryCollector {
+    pub fn new(pid: libc::pid_t) -> Self {
+        let mut descs = Vec::new();
+
+        let vsize = Gauge::with_opts(Opts::new(
+            "process_virtual_memory_bytes",
+            "Virtual memory size in bytes.",
+        ))
+        .unwrap();
+        descs.extend(vsize.desc().into_iter().cloned());
+
+        let rss = Gauge::with_opts(Opts::new(
+            "process_resident_memory_bytes",
+            "Resident memory size in bytes.",
+        ))
+        .unwrap();
+        descs.extend(rss.desc().into_iter().cloned());
+
+        Self {
+            pid,
+            descs,
+            vsize,
+            rss,
+        }
+    }
+}
+
+impl Collector for MemoryCollector {
+    fn desc(&self) -> Vec<&Desc> {
+        self.descs.iter().collect()
+    }
+
+    fn collect(&self) -> Vec<proto::MetricFamily> {
+        if let Ok(statm) = pid_info::statm(self.pid) {
```

Comment:
Is there any special consideration of calling `statm` with pid instead of `statm_self`?

Reply:
nop, it's simply copied from prometheus.